doc/design/containerd/port-layer-integration.md

```diff
@@ -0,0 +1,276 @@
+
+The question of whether snapshots are created as output of image pull or whether these are created as output of container create is one that gets to the heart of whether the output of image pull needs to be cached at all. At present this isn’t clear, but it would be preferable for it to be the latter.
+
+Image metadata will need to persist beyond the lifespan of the VIC appliance and as such will need to be written into vSphere.
+
+*Container Create*
+
+Assuming the image snapshots are already created, an additional RW layer needs to be created, assuming the config doesn’t have readonly=true. 
+
+There is then a configuration pre-processing that needs to happen where the config.json is generated for the container. This will encapsulate a mixture of the image spec, containerd command-line arguments, constants set for the installation and possibly environmental factors.
+
+There will then need to be a validation step of the generated config to make sure that it makes sense. For example, referenced networks and storage need to exist. If the config passes validation, create() is called on the executor interface, which would propagate to the Port Layer. A VMConfigSpec will then be generated and a VM will be created. This is the equivalent of runc create (where a dummy process is created and gets a PID).
+
+*Container Run*
+
+Container run should operate in a very similar manner to the way it does today in that the VM is powered on and a handshake established with the Tether.
+
+The question is how we handle TTY/attach and streaming of stdout and stderr. In the current version of containerd, the containerd-shim maintains a Fifo on the filesystem that can be reconnected to if containerd shuts down and restarts.
+
+<TBD>
+
+**Questions**
+
+There are a number of questions that still need to be worked out. Here’s are the most obvious:
+
+*Do we need to create a process per container in the VIC appliance to fit with the executor model?*
+
+Both runc and the containerd shim are a process-per-container model. Our Port Layer is a long-running service. We will need some kind of shim layer on the Port Layer to mate with the executor interface anyway. Does having entries in /proc for each container help in any way?
+
+*What do we do with the config.json in a bundle?*
+
+When a bundle is extracted to a local filesystem, config.json is used as input to runc in terms of how to run the container. Config.json in the normal case will be a composite of image config and runtime config generated by containerd. For us to integrate seamlessly, we should expect the output of containerd to still be a config.json, but this will then need to be parsed by the Port Layer in creation of the VMConfigSpec. The config.json does not then need to end up in the bundle filesystem mounted to the container since runtime configuration goes into VM guestInfo.
```

Comment:
shim module I am faking has an access to this config, so we could transform it accordingly before talking to portlayer.